v2/handler.js

```diff
@@ -0,0 +1,234 @@
+        self.emit('error', new Error('call request before init request')); // TODO typed error
+        return;
+    }
+
+    var handler = self.channel.getEndpointHandler(name);
+    var headers = {};
+
+    // TODO: couple reqFrame.body.tracing for all downstream requests
+    // TODO: implement service name?
+
+    self.runInOp(handler, {
+        id: id,
+        arg1: reqFrame.body.arg1,
+        arg2: reqFrame.body.arg2,
+        arg3: reqFrame.body.arg3,
+        headers: headers
+    }, function sendResponseFrame(err, res1, res2, callback) {
+        self.sendResponseFrame(reqFrame, headers, err, res1, res2, callback);
+    });
+};
+
+TChannelV2Handler.prototype.handleCallResponse = function handleCallResponse(resFrame) {
+    var self = this;
+    if (self.remoteHostPort === null) {
+        self.emit('error', new Error('call response before init response')); // TODO typed error
+        return;
+    }
+    var id = resFrame.id;
+    var code = resFrame.body.code;
+    var arg1 = resFrame.body.arg1;
+    var arg2 = resFrame.body.arg2;
+    var arg3 = resFrame.body.arg3;
+    if (code === v2.CallResponse.Codes.OK) {
+        self.completeOutOp(null, id, arg2, arg3);
+    } else {
+        self.completeOutOp(TChannelApplicationError({
+            code: code,
+            arg1: arg1,
+            arg2: arg2,
+            arg3: arg3
+        }), id, arg2, null);
+    }
+};
+
+TChannelV2Handler.prototype.handleError = function handleError(errFrame) {
+    var self = this;
+    var id = errFrame.id;
+    var code = errFrame.body.code;
+    var message = errFrame.body.message;
+    var err = v2.ErrorResponse.CodeErrors[code]({message: message});
+    self.completeOutOp(err, id, null, null);
+};
+
+TChannelV2Handler.prototype.sendInitRequest = function sendInitRequest() {
+    var self = this;
+    var id = self.nextFrameId(); // TODO: assert(id === 1)?
+    var hostPort = self.channel.hostPort;
+    var processName = self.channel.processName;
+    var body = v2.InitRequest(v2.VERSION, hostPort, processName);
+    var reqFrame = v2.Frame(id, body);
+    self.writeFrame(reqFrame);
+};
+
+TChannelV2Handler.prototype.sendInitResponse = function sendInitResponse(reqFrame) {
+    var self = this;
+    var id = reqFrame.id;
+    var hostPort = self.channel.hostPort;
+    var processName = self.channel.processName;
+    var body = v2.InitResponse(v2.VERSION, hostPort, processName);
+    var resFrame = v2.Frame(id, body);
+    self.writeFrame(resFrame);
+};
+
+/* jshint maxparams:6 */
+TChannelV2Handler.prototype.sendRequestFrame = function sendRequestFrame(options, arg1, arg2, arg3, callback) {
+    var self = this;
+    var id = self.nextFrameId();
+    var flags = 0; // TODO: streaming
+    var ttl = options.timeout || 1; // TODO: better default, support for dynamic
+    var tracing = options.tracing || null; // TODO: generate
+    var service = options.service || null; // TODO: what even is this? default it to some sort of "my" name?
```

Comment:
oh i see, if you have a ring of tchannels for a single service. I guess you do want the service to be yourself by default.

You also want to check incoming call requests and double check the incoming service is for yourself and otherwise return "i am not a proxy" error.